discovery-infra/test_infra/tools/assets.py

```diff
@@ -1,49 +1,158 @@
-            with open(self._assets_file, "w") as _file:
-                json.dump(all_assets, _file)
-
-    def release_all(self) -> None:
-        logging.info("Returning all %d assets", len(self._took_assets))
-        self._release()
+        with open(self._assets_file) as _file:
+            all_assets = json.load(_file)
+        # move asset to the end of the list
+        asset_data = all_assets.pop(0)
+        all_assets.append(asset_data)
+        with open(self._assets_file, "w") as _file:
+            json.dump(all_assets, _file)
+        logging.info("Taken asset: %s", asset_data)
+        return Munch.fromDict(asset_data)
 
 
 class NetworkAssets(Assets):
+    """ An assets class that stores values based on the current available
+        resources, in order to allow multiple installations while avoiding
+        conflicts. """
+
     def __init__(self):
         super().__init__(assets_file=consts.TF_NETWORK_POOL_PATH)
+        self._allocated_ips_objects = []
+        self._allocated_bridges = []
+
+    def get(self) -> Munch:
+        asset = super().get()
+        self._verify_asset_fields(asset)
+
+        self._allocated_bridges.clear()
+        self._allocated_ips_objects.clear()
+
+        self._find_current_allocated_virsh_net_resources()
+
+        self._override_ip_networks_values_if_not_free(asset)
+        self._override_network_bridges_values_if_not_free(asset)
+
+        logging.info("Using network asset: %s", asset)
+        return asset
+
+    @staticmethod
+    def _verify_asset_fields(asset: Munch):
+        for field in consts.REQUIRED_ASSET_FIELDS:
+            assert hasattr(asset, field), f"missing field {field}"
+
+    def _find_current_allocated_virsh_net_resources(self):
+        for uid in self._virsh_net_resources_uids_generator():
+            net_xml_cmd_output = sp.check_output(["virsh", "net-dumpxml", uid])
+            net_xml = xmltodict.parse(net_xml_cmd_output)
+            self._store_allocated_net_resources_from_net_xml(net_xml)
+
+    @staticmethod
+    def _virsh_net_resources_uids_generator() -> Generator[str, None, None]:
+        cmd_output = sp.check_output(["virsh", "net-list", "--uuid", "--all"])
+        for line in cmd_output.splitlines():
+            uid = line.strip()
+            if not uid:
+                continue
+
+            yield uid
+
+    def _store_allocated_net_resources_from_net_xml(self, net_xml: OrderedDict):
+        try:
+            net_bridge = net_xml["network"]["bridge"]["@name"]
+            self._add_allocated_net_bridge(net_bridge)
+        except KeyError:
+            pass
+
+        try:
+            start_ip = net_xml["network"]["ip"]["dhcp"]["range"]["@start"]
+            end_ip = net_xml["network"]["ip"]["dhcp"]["range"]["@end"]
+            self._add_allocated_ip(IPRange(start_ip, end_ip))
+        except KeyError:
+            pass
+
+    def _override_ip_networks_values_if_not_free(self, asset: Munch):
```

Comment:
why to override ? you should just find one that is free

Reply:
yes and then override the current value of the allocated resource, no?

Reply:
that's the main issue i was talking about. I think that idea is to find free, so you start from some starting point and search for free one, i mean always. In this case logically you are not overriding something but you are searching for free one